File: src/comms/queue.ts
// #EXT-006-REQ-1 Start
/**
 * Rigid, typed queue abstraction — one of the only two sanctioned inter-agent
 * channels (the other being the shared file system, see `./fs`).
 *
 * Agents never address each other directly: a producer `enqueue`s a message and
 * a consumer `dequeue`s it. Every message crosses a schema gate at enqueue time
 * so a contract violation fails loudly with a typed error *before* the message
 * is ever stored or made visible to a consumer.
 *
 * Delivery semantics (explicit, see [REQ-1]):
 *   - Ordering:   in-memory FIFO. Messages are dequeued in the exact order they
 *                 were accepted by `enqueue`.
 *   - Delivery:   at-least-once. A `dequeue` resolves with the next message and
 *                 removes it from the queue. There is no acknowledgement step,
 *                 so a consumer that crashes after `dequeue` but before acting
 *                 simply loses that delivery attempt; producers/consumers must
 *                 tolerate a message being processed one-or-more times. This is
 *                 deliberately NOT exactly-once.
 *   - Durability: in-memory only. The queue does not survive process restart;
 *                 durable exchange is the file system's job (`./fs`).
 *
 * The queue stores no callbacks/handles and exposes no path for a producer to
 * reach a consumer except through the message data itself.
 */

/**
 * A `Validator<T>` decides whether an arbitrary value satisfies a message
 * contract. It is a TypeScript type guard so a successful validation also
 * narrows the value to `T` for the caller.
 *
 * Implementations MUST be pure and MUST throw nothing — they return `false` for
 * any value that violates the contract. The queue turns a `false` result into a
 * {@link QueueContractError}.
 */
export type Validator<T> = (value: unknown) => value is T;

/**
 * Typed error raised when a message fails its contract validation at enqueue
 * time. Contract/layout violations across the comms fabric fail loudly with a
 * typed error ([REQ-4]); this is the queue side of that guarantee.
 */
export class QueueContractError extends Error {
  /** The value that violated the contract (kept for diagnostics). */
  readonly value: unknown;

  constructor(message: string, value: unknown) {
    super(message);
    this.name = "QueueContractError";
    this.value = value;
    // Restore prototype chain for `instanceof` across the CommonJS/ES boundary.
    Object.setPrototypeOf(this, QueueContractError.prototype);
  }
}

/**
 * Generic FIFO queue over a rigid, typed message contract `T`.
 *
 * Construct with a {@link Validator} that defines the contract. Each
 * `enqueue` validates against it and rejects violations with a
 * {@link QueueContractError} before storage.
 */
export class Queue<T> {
  private readonly items: T[] = [];
  private readonly validate: Validator<T>;
  private readonly name: string;

  /**
   * @param validator type guard defining the message contract.
   * @param name      optional queue name used in error messages.
   */
  constructor(validator: Validator<T>, name = "queue") {
    if (typeof validator !== "function") {
      throw new TypeError("Queue requires a validator function.");
    }
    this.validate = validator;
    this.name = name;
  }

  /** Number of messages currently buffered. */
  get size(): number {
    return this.items.length;
  }

  /**
   * Validate `msg` against the contract and, if it passes, append it to the
   * tail of the queue. Throws {@link QueueContractError} on a contract
   * violation; the message is NOT stored in that case.
   */
  enqueue(msg: unknown): void {
    if (!this.validate(msg)) {
      throw new QueueContractError(
        `Message rejected by ${this.name}: does not satisfy the message contract.`,
        msg
      );
    }
    // `validate` narrowed `msg` to `T`.
    this.items.push(msg);
  }

  /**
   * Remove and return the message at the head of the queue (FIFO). Rejects if
   * the queue is empty. At-least-once delivery: the message is removed as it is
   * handed to the caller; there is no acknowledgement.
   */
  async dequeue(): Promise<T> {
    if (this.items.length === 0) {
      throw new Error(`Cannot dequeue from empty ${this.name}.`);
    }
    return this.items.shift() as T;
  }

  /** Non-removing peek at the head message, or `undefined` when empty. */
  peek(): T | undefined {
    return this.items[0];
  }
}
// #EXT-006-REQ-1 End

File: jaros/comms/queue.py
"""Rigid, typed message queue (EXT-006 / REQ-1, REQ-4, REQ-6).

Agents never call each other; they exchange work through queues governed by a
rigid, typed message contract. Every message is schema-validated *before* it is
stored, so a contract violation is rejected at enqueue time and never reaches a
consumer.

This queue is a plain in-process structure: it requires and contacts **no
external message broker or queue service** (EXT-006 / REQ-6) — the comms-layer
half of the zero-infrastructure tenet.

Semantics (explicitly specified):

- **Ordering**: strict FIFO. Messages are dequeued in the order they were
  enqueued.
- **Delivery**: at-least-once. ``dequeue`` removes and returns the head; if a
  consumer crashes after receiving but before durably acting, the message is
  gone from this in-memory queue, but the contract documents at-least-once so
  consumers must be idempotent. (No exactly-once guarantee is offered.)
- **Durability**: none. This is an in-memory queue; contents do not survive
  process restart. Durable exchange is the job of the shared file system
  (``jaros.comms.fs``).
"""

from __future__ import annotations

from collections import deque
from typing import Callable, Generic, TypeVar

T = TypeVar("T")


# #EXT-006-REQ-4 Start
class QueueContractError(ValueError):
    """Raised when a message violates the queue's typed contract at enqueue.

    The offending message is rejected *before* storage, so a violating message
    is never observable by a consumer.
    """
# #EXT-006-REQ-4 End


# #EXT-006-REQ-1 Start
class Queue(Generic[T]):
    """A rigid, typed FIFO queue with enqueue-time contract validation.

    The queue is parameterised by a ``validator`` callable that enforces the
    message contract. ``enqueue`` runs the validator first and raises
    :class:`QueueContractError` on any violation, *before* the message is stored
    — guaranteeing only contract-valid messages are ever delivered.

    Semantics: in-memory, strict FIFO ordering, at-least-once delivery,
    non-durable (see module docstring).
    """

    def __init__(self, validator: Callable[[T], bool] | None = None) -> None:
        """Create a queue.

        Args:
            validator: Callable applied to each message at enqueue time. It may
                either return ``False`` / a falsy value to signal a contract
                violation, or raise its own exception (which is wrapped in a
                :class:`QueueContractError`). If ``None``, all messages are
                accepted.
        """
        self._validator = validator
        self._items: deque[T] = deque()

    def enqueue(self, msg: T) -> None:
        """Validate ``msg`` against the contract, then append it (FIFO tail).

        Raises:
            QueueContractError: if the message violates the contract. The
                message is not stored when this is raised.
        """
        if self._validator is not None:
            try:
                ok = self._validator(msg)
            except QueueContractError:
                raise
            except Exception as exc:  # validator signalled a violation by raising
                raise QueueContractError(
                    f"message rejected by contract validator: {exc}"
                ) from exc
            if not ok:
                raise QueueContractError(
                    f"message rejected by contract validator: {msg!r}"
                )
        self._items.append(msg)

    def dequeue(self) -> T:
        """Remove and return the head of the queue (FIFO).

        Raises:
            IndexError: if the queue is empty.
        """
        if not self._items:
            raise IndexError("dequeue from an empty queue")
        return self._items.popleft()

    def peek(self) -> T:
        """Return the head of the queue without removing it.

        Raises:
            IndexError: if the queue is empty.
        """
        if not self._items:
            raise IndexError("peek from an empty queue")
        return self._items[0]

    def __len__(self) -> int:
        return len(self._items)
# #EXT-006-REQ-1 End
